File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:

    struct cmp {
    bool operator()(ListNode* a, ListNode* b) {
        return a->val > b->val;
    }
};

    ListNode* mergeKLists(vector<ListNode*>& lists) {

        priority_queue<ListNode*, vector<ListNode*>, cmp> pq;

        for (auto list : lists) {
            if (list) {
                pq.push(list);
            }
        }

        if (pq.empty()) {
            return nullptr;
        }

        ListNode* head = pq.top();
        pq.pop();

        ListNode* tail = head;

        if (head->next) {
            pq.push(head->next);
        }

        while (!pq.empty()) {

            ListNode* temp = pq.top();
            pq.pop();

            tail->next = temp;
            tail = temp;

            if (temp->next) {
                pq.push(temp->next);
            }
        }

        return head;
    }
};

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        pq=[]
        count=0

        for list in lists:
            if list:
                heapq.heappush(pq,(list.val,count,list))
                count+=1
        
        if not pq:
            return None

        dummy=ListNode(0)
        tail=dummy

        while pq:
            _,_,node=heapq.heappop(pq)

            tail.next=node
            tail=tail.next

            if node.next:
                heapq.heappush(pq,(node.next.val,count,node.next))
                count+=1

        return dummy.next


        

        